strutil/quantity/quantity.go

```diff
@@ -200,3 +201,262 @@ func FormatDuration(dt float64) string {
+// operation with the remainder below the minimum.
+type RenderMode uint32
+
+const (
+	TimeLeft RenderMode = iota
+	TimePassed
+	TimeRounded
+)
+
+// FormatDurationGeneric formats time duration similar to systemd
+// format_timespan, but with options allowing for a more compact
+// arrangement, including shortened suffixes. The function is provided
+// with dt in seconds (as in the output of time.Now().Seconds()). The 'min'
+// and 'max' Duration allows limiting the unit range rendered. Place values
+// less than 'min' will be processed according to RenderMode. RenderMode
+// allows rounding, flooring or ceiling to be applied to place value below
+// 'min'. Durations above the 'max' place value will result in "ages!".
+func FormatDurationGeneric(dt float64, min Duration, max Duration, count Places, mode RenderMode, space SpaceMode) string {
+	var units = map[Duration]string{
+		Year:    "y",
+		Day:     "d",
+		Hour:    "h",
+		Minute:  "m",
+		Second:  "s",
+		MSecond: "ms",
+		USecond: "µs",
+		NSecond: "ns",
+	}
+
+	// We need to access the map in order using an ordered array
+	// Sort into iteration order: y, d, h, m, s, ...
+	var keys []Duration
+	for k := range units {
+		keys = append(keys, k)
+	}
+	sort.Slice(keys, func(i, j int) bool { return keys[i] > keys[j] })
+
+	var render string
+
+	// Invalid case: min > max
+	if min > max {
+		return "inv!"
+	}
+
+	// Special case: zero duration
+	if dt <= 0 {
+		return "0" + units[min]
+	}
+
+	// Special case: render as "ages!"
+	if dt >= float64(math.MaxUint64) {
+		return "ages!"
+	}
+
+	// Fractional seconds to nanoseconds
+	delta := Duration(dt * float64(Second))
+
+	// If we only render a subset of place values, we need to apply
+	// the render mode to the last place value. We do this indirectly
+	// by moving up the 'min' accuracy.
+	if count == ShowCompact {
+		for i, key := range keys {
+
+			// The duration is smaller than this place value, skip
+			if delta < key {
+				continue
+			}
+
+			// Compact mode only renders two places, up until seconds
+			// The indexing is safe as places beyond seconds exist
+			if key >= Minute {
+				index := i + 1
+				if keys[index] > min {
+					min = keys[index]
+				}
+				break
+			}
+
+			// Only one place if we are s, ms, us or ns
+			if key < Minute {
+				if key > min {
+					min = key
+				}
+				break
+			}
+		}
+	}
+
+	// Apply the rendermode to the least significant place value
+	mod := delta % min
+	switch mode {
+	case TimeLeft:
+		// Ceiling
+		if mod > 0 {
+			delta += min - mod
+		}
+	case TimePassed:
+		// Floor
+		delta -= mod
+	case TimeRounded:
+		if mod >= (min / 2) {
+			// Rounded Up
+			delta += min - mod
+		} else {
+			// Rounded Down
+			delta -= mod
+		}
+	default:
+		// Invalid mode
+		return "inv!"
+	}
+
+	// Special case: less than minimum accuracy
+	if delta < min {
+		return "0" + units[min]
+	}
+
+	// In ShowCompact mode only 2-digit days are possible
+	if count == ShowCompact && delta >= (100*Day) {
```

Comment:
but what if `delta` is 2 years? Here you would be returning `ages!`, while you should return `2y`, unless I'm mistaken.

Reply:
Compact Mode is a very special mode where the space requirements outweighs the accuracy and range. Compact Mode was designed for User Interactive Progress indicators, where a user will wait for an operation to complete, and may want to look at the (e.g. ANSIMeter) output for a hint how long to still wait for, or if something is wrong (i.e. Ethernet is basically broken and the package update is getting close to 0 bytes of data through). 

The function using Compact Mode actually sets the minimum duration to Hours, so even Days are excluded for ANSIMeter. 2 Digit Days (0-99) is the best the Compact Mode can offer for User Interfaces that really expect a progress meter with Day level units. This is tested, but this is not currently used by a use-case.

Reply:
I see. But then I think I'm going to ask another thing: since it seems that we have some preconditions on how this function's parameters constrain each other, we should not really rely on the caller doing the right thing. What I would suggest to do is to validate the parameters at the beginning of the function, for example with code like:
```
if count == ShowCompact && mode == TimeLeft && max > Hour {
     log.Noticef("Invalid value %v for max duration, limiting to Hour", max)
     max = Hour
}
```
and so on (the above is just an example, you know the logic better than me). In this way all this constraints will be documented, and I hopefully won't bug you with similar requests.

Reply:
Discussed. Compact Mode changed to be more consistent, so it can not render years. However, since years and days could be more than 3 runes (the allowed maximum for two places to be rendered), some cases will reduce to a single place value. See the example_test.go for the two actual use cases.

Fixed.